homeassistant/components/tolo/sensor.py

```diff
@@ -0,0 +1,73 @@
+"""TOLO Sauna (non-binary, general) sensors."""
+
+from homeassistant.components.sensor import SensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ENTITY_CATEGORY_DIAGNOSTIC, PERCENTAGE, TEMP_CELSIUS
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ToloSaunaCoordinatorEntity, ToloSaunaUpdateCoordinator
+from .const import DOMAIN
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up (non-binary, general) sensors for TOLO Sauna."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
+            ToloWaterLevelSensor(coordinator, entry),
+            ToloTankTemperatureSensor(coordinator, entry),
+        ]
+    )
+
+
+class ToloWaterLevelSensor(ToloSaunaCoordinatorEntity, SensorEntity):
+    """Sensor for tank water level."""
+
+    WATER_LEVELS = {0: 0, 1: 33, 2: 66, 3: 100}
```

Comment:
I think this is faking the levels a bit? As in, the percentage isn't its native value, but you are trying to resolve the arbitrary number?

Reply:
Exactly. Level is represented by the device display with a 3 bars symbol, so 4 different steps, and network protocol returns int from 0 to 3, whereby 0 is empty and 3 is full.

Is there any better way to solve this?

Reply:
I personally would present it as, and add the device class `tolo__water_level` as the `_attr_device_class`.

Next, make a file `strings.sensor.json` in your integration folder with:

```json
{
  "state": {
    "tuya__status": {
      "0": "Empty",
      "1": "Low",
      "2": "Normal",
      "3": "Full"
    }
  }
}
```

(or something like that, whatever fits better).

Next run `python -m script.translations develop` to update the translations with this.

This makes the statuses translatable and look nice in the UI and doesn't fake it :)

More information: <https://developers.home-assistant.io/docs/internationalization/core/#stringssensorjson>

Reply:
Thx for the nice example! I tried to implement it that way, but still have no better word for `33 %` or `66 %`. Normal operation of the sauna is with Full tank, "low/high" might be misleading, so the most neutral and non-interpreting thing of the values `1` and `2` is `33 %` and `66 %`.

Reply:
I agree normal and high sound wierd, maybe low/mid? (Empty, Low, Mid, Full)
I think absolute values are weird in this case a bit, because it isn't an absolute value :)

Reply:
I'm really not sure what the best way here is. "Mid" leads to interpretation of the actual status, which could be understood to be "50%". However, according to a verbally communicated information of the manufacturer, 0..3 directly represent the level from 0% to 100%...

Also, with the previous solution with returning directly the percentages, HA plotted a nice graph about the filling state - which is gone now of course. During operations, when a lot of steam was generated, water level can shortly fall from 3 to 2, which then will result in opening the water in valve (will come with my next pull request, `binary_sensor` :D) and then the value will go back to 3. Having this plotted in a history graph would be quite nice in my opinion...

I would prefer going back to the (maybe a bit misused and generously interpreted meaning of the word "native") solution with returning 0/33/66/100 from the `native_value` property, because it natively represents the water level, just the network protocol is not using 0/33/66/100 but 0..3.

You're the boss, I leave it up to your decision.

Reply:
> You're the boss, I leave it up to your decision.

Nah I'm not. Just trying to think along and discuss and see if we got something nice out of it.

Anyways, you are right, the values are nice there too for graphing. Maybe the library could do the translations and provide Home Assistant with the percentage?

Reply:
Done :) However, this now requires a dependency update from `tololib==0.1.0b2` to `tololib==0.1.0b3` (only non-breaking changes, see https://gitlab.com/MatthiasLohr/tololib/-/releases/v0.1.0b3).